pkg/networkservice/common/heal/client.go

```diff
@@ -21,250 +19,41 @@ package heal
 
-	"github.com/edwarnicke/serialize"
-
-	"github.com/networkservicemesh/sdk/pkg/tools/addressof"
-
 	"github.com/networkservicemesh/sdk/pkg/networkservice/core/next"
 )
 
+// RegisterClientFunc - required to inform heal server about new client connection and assign it to the connection ID
+type RegisterClientFunc func(context.Context, *networkservice.Connection, networkservice.MonitorConnectionClient)
+
 type healClient struct {
-	ctx                   context.Context
-	client                networkservice.MonitorConnectionClient
-	onHeal                *networkservice.NetworkServiceClient
-	cancelHealMap         map[string]func() <-chan error
-	cancelHealMapExecutor serialize.Executor
+	ctx            context.Context
+	cc             networkservice.MonitorConnectionClient
+	registerClient RegisterClientFunc
 }
 
-// NewClient - creates a new networkservice.NetworkServiceClient chain element that implements the healing algorithm
-//             - ctx    - context for the lifecycle of the *Client* itself.  Cancel when discarding the client.
-//             - client - networkservice.MonitorConnectionClient that can be used to call MonitorConnection against the endpoint
-//             - onHeal - *networkservice.NetworkServiceClient.  Since networkservice.NetworkServiceClient is an interface
-//                        (and thus a pointer) *networkservice.NetworkServiceClient is a double pointer.  Meaning it
-//                        points to a place that points to a place that implements networkservice.NetworkServiceClient
-//                        This is done because when we use heal.NewClient as part of a chain, we may not *have*
-//                        a pointer to this
-//                        client used 'onHeal'.  If we detect we need to heal, onHeal.Request is used to heal.
-//                        If onHeal is nil, then we simply set onHeal to this client chain element
-//                        If we are part of a larger chain or a server, we should pass the resulting chain into
-//                        this constructor before we actually have a pointer to it.
-//                        If onHeal nil, onHeal will be pointed to the returned networkservice.NetworkServiceClient
-func NewClient(ctx context.Context, client networkservice.MonitorConnectionClient, onHeal *networkservice.NetworkServiceClient) networkservice.NetworkServiceClient {
-	rv := &healClient{
-		ctx:           ctx,
-		client:        client,
-		onHeal:        onHeal,
-		cancelHealMap: make(map[string]func() <-chan error),
+// NewClient - creates a new networkservice.NetworkServiceClient chain element that inform healServer about new client connection
+func NewClient(ctx context.Context, cc networkservice.MonitorConnectionClient, registerClient RegisterClientFunc) networkservice.NetworkServiceClient {
+	return &healClient{
+		ctx:            ctx,
+		cc:             cc,
+		registerClient: registerClient,
```

Comment:
This parameter can be removed if we go with the suggestion above.

Reply:
Done